x-pack/plugin/monitoring/src/internalClusterTest/java/org/elasticsearch/xpack/monitoring/integration/MonitoringIT.java

```diff
@@ -206,27 +213,39 @@ public void testMonitoringService() throws Exception {
         assertAcked(clusterAdmin().prepareUpdateSettings().setTransientSettings(settings));
 
         whenExportersAreReady(() -> {
-            final AtomicReference<SearchResponse> searchResponse = new AtomicReference<>();
-
-            assertBusy(() -> {
-                final SearchResponse response = client().prepareSearch(".monitoring-es-*")
-                    .setCollapse(new CollapseBuilder("type"))
-                    .addSort("timestamp", SortOrder.DESC)
-                    .get();
-
-                assertThat(response.status(), is(RestStatus.OK));
-                assertThat(
-                    "Expecting a minimum number of 6 docs, one per collector",
-                    response.getHits().getHits().length,
-                    greaterThanOrEqualTo(6)
-                );
-
-                searchResponse.set(response);
-            });
+            final AtomicReference<SearchResponse> searchResponseRef = new AtomicReference<>();
+            try {
+                assertBusy(() -> {
+                    final SearchResponse response = client().prepareSearch(".monitoring-es-*")
+                        .setCollapse(new CollapseBuilder("type"))
+                        .addSort("timestamp", SortOrder.DESC)
+                        .get();
+
+                    try {
+                        assertThat(response.status(), is(RestStatus.OK));
+                        assertThat(
+                            "Expecting a minimum number of 6 docs, one per collector",
+                            response.getHits().getHits().length,
+                            greaterThanOrEqualTo(6)
+                        );
+                    } catch (Exception e) {
+                        // if an exception is thrown, then we will not set the AtomicReference
+                        // so we need to decRef this response that is failing the asserts
+                        response.decRef();
+                        throw e;
+                    }
 
-            for (final SearchHit hit : searchResponse.get().getHits()) {
-                final Map<String, Object> searchHit = toMap(hit);
-                assertMonitoringDoc(searchHit, MonitoredSystem.ES, MonitoringService.MIN_INTERVAL);
+                    searchResponseRef.set(response);
```

Comment:
Wouldn't we have to use `setAndGet` here and release whatever non-null response that might already be in the reference ?

Reply:
We could do that for safety, but I don't think it's necessary (?). The AtomicReference set called should only be called once, since the assertBusy "loop" stops only after none of the asserts thrown an Exception. I catch the Exception above and decRef those failed SearchResponses and then re-throw, so the setter should only be called once when the assertBusy finally returns without error.

So I could add your idea for safety or put in an assert that the searchResponseRef is null before setting it.

Reply:
I don't understand what we need all of this. I think this should be enough:

```
        whenExportersAreReady(() -> {
            assertBusy(() -> {
                assertCheckedResponse(client().prepareSearch(".monitoring-es-*")
                        .setCollapse(new CollapseBuilder("type"))
                        .addSort("timestamp", SortOrder.DESC)
                    , response -> {
                        assertThat(response.status(), is(RestStatus.OK));
                        assertThat(
                            "Expecting a minimum number of 6 docs, one per collector",
                            response.getHits().getHits().length,
                            greaterThanOrEqualTo(6)
                        );

                        for (final SearchHit hit : response.getHits()) {
                            final Map<String, Object> searchHit = toMap(hit);
                            assertMonitoringDoc(searchHit, MonitoredSystem.ES, MonitoringService.MIN_INTERVAL);
                        }
                    });
            });
        });
```